src/main/java/com/aws/greengrass/mqttclient/MqttClient.java

```diff
@@ -225,67 +231,76 @@ protected MqttClient(DeviceConfiguration deviceConfiguration,
+            Future<?> oldFuture = reconfigureFuture.getAndSet(ses.schedule(() -> {
+                // If the rootCa path changed, then we need to update the TLS options
+                String newRootCaPath = Coerce.toString(deviceConfiguration.getRootCAFilePath());
+                synchronized (this) {
+                    if (!Objects.equals(rootCaPath, newRootCaPath)) {
+                        if (proxyTlsOptions != null) {
+                            proxyTlsOptions.close();
+                        }
+                        if (proxyTlsContext != null) {
+                            proxyTlsContext.close();
+                        }
+                        rootCaPath = newRootCaPath;
+                        proxyTlsOptions = getTlsContextOptions(rootCaPath);
+                        proxyTlsContext = new ClientTlsContext(proxyTlsOptions);
+                    }
                 }
 
-                logger.atDebug().kv("modifiedNode", node.getFullName()).kv("changeType", what)
-                        .log("Reconfiguring MQTT clients");
-
-                // Reconnect in separate thread to not block publish thread
-                // Schedule the reconnection for slightly in the future to de-dupe multiple changes
-                Future<?> oldFuture = reconfigureFuture.getAndSet(ses.schedule(() -> {
-                    // Continually try to reconnect until all the connections are reconnected
-                    Set<AwsIotMqttClient> brokenConnections = new CopyOnWriteArraySet<>(connections);
-                    do {
-                        for (AwsIotMqttClient connection : brokenConnections) {
-                            if (Thread.currentThread().isInterrupted()) {
-                                return;
-                            }
+                // Continually try to reconnect until all the connections are reconnected
+                Set<AwsIotMqttClient> brokenConnections = new CopyOnWriteArraySet<>(connections);
+                do {
+                    for (AwsIotMqttClient connection : brokenConnections) {
+                        if (Thread.currentThread().isInterrupted()) {
+                            return;
+                        }
 
-                            try {
-                                connection.reconnect();
-                                brokenConnections.remove(connection);
-                            } catch (InterruptedException | ExecutionException | TimeoutException e) {
-                                logger.atError().setCause(e).kv(CLIENT_ID_KEY, connection.getClientId())
-                                        .log("Error while reconnecting MQTT client");
-                            }
+                        try {
+                            connection.reconnect();
+                            brokenConnections.remove(connection);
+                        } catch (InterruptedException | ExecutionException | TimeoutException e) {
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

**Problem**: InterruptedException is ignored. This can delay thread shutdown and clear the thread’s interrupt status. Only code that implements a thread’s interruption policy can swallow an interruption request.

**Fix**: Rethrow the InterruptedException or reinterrupt the current thread using *Thread.currentThread().interrupt()* so that higher-level interrupt handlers can function correctly.
 If you are wrapping the InterruptedException inside a RuntimeException, call *Thread.currentThread().interrupt()* before throwing the RuntimeException.

Learn more about [interrupts](https://docs.oracle.com/javase/tutorial/essential/concurrency/interrupt.html) and [dealing with InterruptedException](https://www.ibm.com/developerworks/java/library/j-jtp05236/index.html?ca=drs-#N10187)